src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,501 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+
+            ssx::spawn_with_gate(_workers_gate, [&]() mutable {
+                return log_and_ignore_exceptions(start_shard_fetch_worker(
+                  shard, octx, std::move(plan.fetches_per_shard[shard]), 0));
+            });
+        }
+
+        for (;;) {
+            co_await progress_conditions(octx);
+
+            if (octx.should_stop_fetch()) {
+                // Send abort signal to workers and wait for all workers to end
+                // here.
+                co_await _as.request_abort();
+                co_await _workers_gate.close();
+                co_await _as.stop();
+                co_return;
+            }
+
+            std::vector<shard_fetch> completed_shard_fetches = std::move(
+              _completed_shard_fetches);
+            _completed_shard_fetches.clear();
+
+            for (auto& sf : completed_shard_fetches) {
+                auto shard = sf.shard;
+                ssx::spawn_with_gate(
+                  _workers_gate,
+                  [this, &octx, shard, sf = std::move(sf)]() mutable {
+                      return log_and_ignore_exceptions(start_shard_fetch_worker(
+                        shard,
+                        octx,
+                        std::move(sf),
+                        // Require that a worker returns more data than before.
+                        // Otherwise it'll return right away with the previous
+                        // result.
+                        _last_result_size[shard] + 1));
+                  });
+            }
+        }
+    }
+
+private:
+    /**
+     * Determines if the should_stop_fetch() condition should be checked again.
+     * The return future is set if;
+     * - octx.deadline has been reached.
+     * - _as has been aborted.
+     * - one of the shard workers has returned results.
+     */
+    ss::future<> progress_conditions(op_context& octx) {
+        ss::promise<> timeout_pr, abort_pr;
+
+        auto abort_sub_opt = _as.subscribe(
+          [&abort_pr]() noexcept { abort_pr.set_value(); });
+
+        if (!abort_sub_opt) {
+            abort_pr.set_value();
+        }
+
+        ss::timer<model::timeout_clock> s{
+          [&timeout_pr] { timeout_pr.set_value(); }};
+
+        if (octx.deadline) {
+            s.arm(octx.deadline.value());
+        }
+
+        co_await ss::when_any(
+          ignore_exceptions(_has_data.wait()),
+          ignore_exceptions(timeout_pr.get_future()),
+          ignore_exceptions(abort_pr.get_future()));
+    }
+
+    /*
+     * `start_shard_fetch_worker` executes on the coordinator shard. It builds
+     * the `shard_local_fetch_context` struct needed to start the
+     * shard_fetch_worker then makes a cross shard call to the shard fetch
+     * worker if needed.
+     *
+     * It then waits until the shard_fetch_worker completes
+     * and then notifies the fetch coordinator.
+     */
+    ss::future<> start_shard_fetch_worker(
+      ss::shard_id shard,
```

Comment:
isn't this redundant with the id in the `fetch` object passed as the 3rd param?

Reply:
Right you are, removing the `shard` parameter.